### ISSUE TITLE
FW: perl 5.6.0 -> perl 5.6.1 XS &AUTOLOAD behaviour change.

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#3869]
Migrated from [rt.perl.org#6828](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=6828) (status was 'resolved')

Searchable as RT6828$

**Comment by p5pRT** [comment]
### From gavin.brock@nssmb.com

Hi\,

I think this a bug\, or at least an implementation change\. We have a module\,
linked against a C library\, that uses an AUTOLOAD function implemented in
XS\. This worked fine up to 5\.6\.0\, but under 5\.6\.1\, our C AUTOLOAD function
is no longer called when an undeclared function is requested\.

We have enclosed\, a minimal module that implements just AUTOLOAD and
DESTROY\. The 'make test' output for each perl version is shown below\. 


>>> make test under 5\.6\.0 \<\<\<

PERL\_DL\_NONLAZY=1 /opt/perl5\.6/bin/perl \-Iblib/arch \-Iblib/lib
\-I/opt/perl5\.6/lib/5\.6\.0/sun4\-solaris\-multi\-64int \-I/opt/perl5\.6/lib/5\.6\.0
test\.pl
\(FALIP561\.cc&#8203;:71\) FALIP561&#8203;::new
\(FALIP561\.cc&#8203;:134\) FALIP561&#8203;::AUTOLOAD
\(FALIP561\.cc&#8203;:153\) hello
\(FALIP561\.cc&#8203;:105\) FALIP561&#8203;::DESTROY


>>> make test under 5\.6\.1 \<\<\<

PERL\_DL\_NONLAZY=1 /opt/perl5\.6/bin/perl \-Iblib/arch \-Iblib/lib
\-I/opt/perl5\.6/lib/5\.6/sun4\-solaris\-multi\-64int \-I/opt/perl5\.6/lib/5\.6
test\.pl
\(FALIP561\.cc&#8203;:71\) FALIP561&#8203;::new
Can't locate object method "hello" via package "FALIP561" \(perhaps you
forgot to load "FALIP561"?\) at test\.pl line 10\.
\(FALIP561\.cc&#8203;:105\) FALIP561&#8203;::DESTROY
\*\*\* Error code 255
make&#8203;: Fatal error&#8203;: Command failed for target \`test\_dynamic'


Is this a bug\, or is our declaration of AUTOLOAD bad?

Thanks in advance\,


Gavin Brock
Jonathon Fletcher

**Comment by p5pRT** [comment]
### From gavin.brock@nssmb.com

[test.pl](https://rt-archive.perl.org/perl5/Ticket/Attachment/16581/17205/test.pl)

**Comment by p5pRT** [comment]
### From gavin.brock@nssmb.com

[FALIP561.pm](https://rt-archive.perl.org/perl5/Ticket/Attachment/16581/17206/FALIP561.pm)

**Comment by p5pRT** [comment]
### From gavin.brock@nssmb.com

[FALIP561.cc](https://rt-archive.perl.org/perl5/Ticket/Attachment/16581/17207/FALIP561.cc)

**Comment by p5pRT** [comment]
### From gavin.brock@nssmb.com

[Makefile.PL](https://rt-archive.perl.org/perl5/Ticket/Attachment/16581/17208/Makefile.PL)

**Comment by p5pRT** [comment]
### From gavin.brock@nssmb.com

This is a bug report for perl from gavin\.brock@&#8203;nssmb\.com & jonathon\.fletcher@&#8203;nssmb\.com\,
generated with the help of perlbug 1\.28 running under perl v5\.6\.0\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl v5.6.0:

Configured by jonathon at Wed Aug  2 10:32:58 JST 2000.

Summary of my perl5 (revision 5.0 version 6 subversion 0) configuration:
  Platform:
    osname=solaris, osvers=2.5.1, archname=sun4-solaris-multi-64int
    uname='sunos pueblo 5.5.1 generic_103640-31 sun4u sparc sunw,ultra-2 '
    config_args=''
    hint=previous, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=define
    useperlio=undef d_sfio=undef uselargefiles=define 
    use64bitint=undef use64bitall=undef uselongdouble=undef usesocks=undef
  Compiler:
    cc='gcc', optimize='-O', gccversion=2.95.2 19991024 (release)
    cppflags='-DPACK_MALLOC -DTWO_POT_OPTIMIZE -fno-strict-aliasing -I/opt/gnu/include -I/usr/local/include'
    ccflags ='-DPACK_MALLOC -DTWO_POT_OPTIMIZE -fno-strict-aliasing -I/opt/gnu/include -I/usr/local/include'
    stdchar='unsigned char', d_stdstdio=define, usevfork=false
    intsize=4, longsize=4, ptrsize=4, doublesize=8
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=4
    alignbytes=8, usemymalloc=y, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags ='-L/opt/gnu/lib -R/opt/gnu/lib -L/usr/local/lib -R/usr/local/lib '
    libpth=/lib /usr/lib /usr/ccs/lib /opt/gnu/lib /usr/local/lib
    libs=-lsocket -lnsl -ldb -ldl -lm -lc -lcrypt -lsec
    libc=/lib/libc.so, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags=' '
    cccdlflags='-fPIC', lddlflags='-G -L/opt/gnu/lib -R/opt/gnu/lib -L/usr/local/lib -R/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.6.0:
    /opt/perl5.6/lib/5.6.0/sun4-solaris-multi-64int
    /opt/perl5.6/lib/5.6.0
    /opt/perl5.6/lib/site_perl/5.6.0/sun4-solaris-multi-64int
    /opt/perl5.6/lib/site_perl/5.6.0
    /opt/perl5.6/lib/site_perl
    .


Environment for perl v5.6.0:
    HOME=/
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/sbin:/usr/sbin:/usr/bin:/usr/ccs/bin:/opt/bin:/opt/bin:/usr/local/bin
    PERL_BADLANG (unset)
    SHELL=/sbin/sh

```
</details>

**Comment by p5pRT** [comment]
### From gavin.brock@nssmb.com

This is a bug report for perl from gavin\.brock@&#8203;nssmb\.com & jonathon\.fletcher@&#8203;nssmb\.com\,
generated with the help of perlbug 1\.28 running under perl v5\.6\.0\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl v5.6.0:

Configured by jonathon at Wed Aug  2 10:32:58 JST 2000.

Summary of my perl5 (revision 5.0 version 6 subversion 0) configuration:
  Platform:
    osname=solaris, osvers=2.5.1, archname=sun4-solaris-multi-64int
    uname='sunos pueblo 5.5.1 generic_103640-31 sun4u sparc sunw,ultra-2 '
    config_args=''
    hint=previous, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=define
    useperlio=undef d_sfio=undef uselargefiles=define 
    use64bitint=undef use64bitall=undef uselongdouble=undef usesocks=undef
  Compiler:
    cc='gcc', optimize='-O', gccversion=2.95.2 19991024 (release)
    cppflags='-DPACK_MALLOC -DTWO_POT_OPTIMIZE -fno-strict-aliasing -I/opt/gnu/include -I/usr/local/include'
    ccflags ='-DPACK_MALLOC -DTWO_POT_OPTIMIZE -fno-strict-aliasing -I/opt/gnu/include -I/usr/local/include'
    stdchar='unsigned char', d_stdstdio=define, usevfork=false
    intsize=4, longsize=4, ptrsize=4, doublesize=8
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=4
    alignbytes=8, usemymalloc=y, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags ='-L/opt/gnu/lib -R/opt/gnu/lib -L/usr/local/lib -R/usr/local/lib '
    libpth=/lib /usr/lib /usr/ccs/lib /opt/gnu/lib /usr/local/lib
    libs=-lsocket -lnsl -ldb -ldl -lm -lc -lcrypt -lsec
    libc=/lib/libc.so, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags=' '
    cccdlflags='-fPIC', lddlflags='-G -L/opt/gnu/lib -R/opt/gnu/lib -L/usr/local/lib -R/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.6.0:
    /opt/perl5.6/lib/5.6.0/sun4-solaris-multi-64int
    /opt/perl5.6/lib/5.6.0
    /opt/perl5.6/lib/site_perl/5.6.0/sun4-solaris-multi-64int
    /opt/perl5.6/lib/site_perl/5.6.0
    /opt/perl5.6/lib/site_perl
    .


Environment for perl v5.6.0:
    HOME=/
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/sbin:/usr/sbin:/usr/bin:/usr/ccs/bin:/opt/bin:/opt/bin:/usr/local/bin
    PERL_BADLANG (unset)
    SHELL=/sbin/sh

```
</details>

**Comment by p5pRT** [comment]
### From gavin.brock@nssmb.com

#### Created by gavin.brock@nssmb.com



\[NOTE&#8203;: This a resend of bug id 20010416\.001 which didn't seem to make it
into the system properly\]

Hi\,

I think this a bug\, or at least an implementation change\. We have a module\,
linked against a C library\, that uses an AUTOLOAD function implemented in
XS\. This worked fine up to 5\.6\.0\, but under 5\.6\.1\, our C AUTOLOAD function
is no longer called when an undeclared function is requested\.

We have enclosed\, a minimal module that implements just AUTOLOAD and
DESTROY\. The 'make test' output for each perl version is shown below\. 


>>> make test under 5\.6\.0 \<\<\<

PERL\_DL\_NONLAZY=1 /opt/perl5\.6/bin/perl \-Iblib/arch \-Iblib/lib
\-I/opt/perl5\.6/lib/5\.6\.0/sun4\-solaris\-multi\-64int \-I/opt/perl5\.6/lib/5\.6\.0
test\.pl
\(FALIP561\.cc&#8203;:71\) FALIP561&#8203;::new
\(FALIP561\.cc&#8203;:134\) FALIP561&#8203;::AUTOLOAD
\(FALIP561\.cc&#8203;:153\) hello
\(FALIP561\.cc&#8203;:105\) FALIP561&#8203;::DESTROY


>>> make test under 5\.6\.1 \<\<\<

PERL\_DL\_NONLAZY=1 /opt/perl5\.6/bin/perl \-Iblib/arch \-Iblib/lib
\-I/opt/perl5\.6/lib/5\.6/sun4\-solaris\-multi\-64int \-I/opt/perl5\.6/lib/5\.6
test\.pl
\(FALIP561\.cc&#8203;:71\) FALIP561&#8203;::new
Can't locate object method "hello" via package "FALIP561" \(perhaps you
forgot to load "FALIP561"?\) at test\.pl line 10\.
\(FALIP561\.cc&#8203;:105\) FALIP561&#8203;::DESTROY
\*\*\* Error code 255
make&#8203;: Fatal error&#8203;: Command failed for target \`test\_dynamic'


Is this a bug\, or is our declaration of AUTOLOAD bad?

Thanks in advance\,


Gavin Brock
Jonathon Fletcher

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl v5.6.1:

Configured by jonathon at Fri Apr 13 14:23:47 JST 2001.

Summary of my perl5 (revision 5.0 version 6 subversion 1) configuration:
  Platform:
    osname=solaris, osvers=2.5.1, archname=sun4-solaris-multi-64int
    uname='sunos pueblo 5.5.1 generic_103640-31 sun4u sparc sunw,ultra-2 '
    config_args=''
    hint=previous, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef
useithreads'perlbug@perl.com'=undef usemultiplicity=define
    useperlio=undef d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
  Compiler:
    cc='gcc', ccflags ='-DPACK_MALLOC -DTWO_POT_OPTIMIZE
-fno-strict-aliasing -I/opt/gnu/include -I/usr/local/include',
    optimize='-O',
    cppflags='-DPACK_MALLOC -DTWO_POT_OPTIMIZE -fno-strict-aliasing
-I/opt/gnu/include -I/usr/local/include'
    ccversion='', gccversion='2.95.2 19991024 (release)',
gccosandvers='solaris2.5.1'
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=4321
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t',
lseeksize=4
    alignbytes=8, usemymalloc=y, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags ='-L/opt/gnu/lib -R/opt/gnu/lib -L/usr/local/lib
-R/usr/local/lib '
    libpth=/lib /usr/lib /usr/ccs/lib /opt/gnu/lib /usr/local/lib
    libs=-lsocket -lnsl -ldb -ldl -lm -lc -lcrypt -lsec
    perllibs=-lsocket -lnsl -ldl -lm -lc -lcrypt -lsec
    libc=/lib/libc.so, so=so, useshrplib=true, libperl=libperl.so
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='  -R
/opt/perl5.6/lib/5.6/sun4-solaris-multi-64int/CORE'
    cccdlflags='-fPIC', lddlflags='-G -L/opt/gnu/lib -R/opt/gnu/lib
-L/usr/local/lib -R/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.6.1:
    /home/user/gb83196/perl/lib/
    /opt/perl5.6/lib/5.6/sun4-solaris-multi-64int
    /opt/perl5.6/lib/5.6
    /opt/perl5.6/lib/site_perl/5.6/sun4-solaris-multi-64int
    /opt/perl5.6/lib/site_perl/5.6
    /opt/perl5.6/lib/site_perl/5.6
    .


Environment for perl v5.6.1:
    HOME=/home/user/gb83196
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH=/home/user/gb83196/rvplayer:/usr/dt/lib
    LOGDIR (unset)
 
PATH=/home/user/gb83196/bin:/home/user/gb83196/private-bin:/opt/perl5.6/bin:
/home/user/gb83196/bin:/home/sb/teknekron/tss_root/rel_s4_5x/bin:/usr/dt/bin
:/usr/openwin/bin:/usr/bin:/usr/ccs/bin:/usr/ucb:/usr/local/bin:/usr/share/l
ocal/bin:/opt/bin:.:/usr/sbin:/xenv/java/sun4/5.x/1.1.3/bin:/home/user/gb831
96/rvplayer
    PERLLIB=/home/user/gb83196/perl/lib/
    PERL_BADLANG (unset)
    SHELL=/home/user/gb83196/bin/sun5//tcsh


```
</details>

**Comment by p5pRT** [comment]
### From gavin.brock@nssmb.com

[test.pl](https://rt-archive.perl.org/perl5/Ticket/Attachment/16920/17568/test.pl)

**Comment by p5pRT** [comment]
### From gavin.brock@nssmb.com

[FALIP561.pm](https://rt-archive.perl.org/perl5/Ticket/Attachment/16920/17569/FALIP561.pm)

**Comment by p5pRT** [comment]
### From gavin.brock@nssmb.com

[FALIP561.cc](https://rt-archive.perl.org/perl5/Ticket/Attachment/16920/17570/FALIP561.cc)

**Comment by p5pRT** [comment]
### From gavin.brock@nssmb.com

[Makefile.PL](https://rt-archive.perl.org/perl5/Ticket/Attachment/16920/17571/Makefile.PL)

**Comment by p5pRT** [comment]
### From gavin.brock@nssmb.com

This is a bug report for perl from gavin\.brock@&#8203;nssmb\.com & jonathon\.fletcher@&#8203;nssmb\.com\,
generated with the help of perlbug 1\.28 running under perl v5\.6\.0\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl v5.6.0:

Configured by jonathon at Wed Aug  2 10:32:58 JST 2000.

Summary of my perl5 (revision 5.0 version 6 subversion 0) configuration:
  Platform:
    osname=solaris, osvers=2.5.1, archname=sun4-solaris-multi-64int
    uname='sunos pueblo 5.5.1 generic_103640-31 sun4u sparc sunw,ultra-2 '
    config_args=''
    hint=previous, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=define
    useperlio=undef d_sfio=undef uselargefiles=define 
    use64bitint=undef use64bitall=undef uselongdouble=undef usesocks=undef
  Compiler:
    cc='gcc', optimize='-O', gccversion=2.95.2 19991024 (release)
    cppflags='-DPACK_MALLOC -DTWO_POT_OPTIMIZE -fno-strict-aliasing -I/opt/gnu/include -I/usr/local/include'
    ccflags ='-DPACK_MALLOC -DTWO_POT_OPTIMIZE -fno-strict-aliasing -I/opt/gnu/include -I/usr/local/include'
    stdchar='unsigned char', d_stdstdio=define, usevfork=false
    intsize=4, longsize=4, ptrsize=4, doublesize=8
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=4
    alignbytes=8, usemymalloc=y, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags ='-L/opt/gnu/lib -R/opt/gnu/lib -L/usr/local/lib -R/usr/local/lib '
    libpth=/lib /usr/lib /usr/ccs/lib /opt/gnu/lib /usr/local/lib
    libs=-lsocket -lnsl -ldb -ldl -lm -lc -lcrypt -lsec
    libc=/lib/libc.so, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags=' '
    cccdlflags='-fPIC', lddlflags='-G -L/opt/gnu/lib -R/opt/gnu/lib -L/usr/local/lib -R/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.6.0:
    /opt/perl5.6/lib/5.6.0/sun4-solaris-multi-64int
    /opt/perl5.6/lib/5.6.0
    /opt/perl5.6/lib/site_perl/5.6.0/sun4-solaris-multi-64int
    /opt/perl5.6/lib/site_perl/5.6.0
    /opt/perl5.6/lib/site_perl
    .


Environment for perl v5.6.0:
    HOME=/
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/sbin:/usr/sbin:/usr/bin:/usr/ccs/bin:/opt/bin:/opt/bin:/usr/local/bin
    PERL_BADLANG (unset)
    SHELL=/sbin/sh

```
</details>

**Comment by p5pRT** [comment]
### From gavin.brock@nssmb.com

This is a bug report for perl from gavin\.brock@&#8203;nssmb\.com & jonathon\.fletcher@&#8203;nssmb\.com\,
generated with the help of perlbug 1\.28 running under perl v5\.6\.0\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl v5.6.0:

Configured by jonathon at Wed Aug  2 10:32:58 JST 2000.

Summary of my perl5 (revision 5.0 version 6 subversion 0) configuration:
  Platform:
    osname=solaris, osvers=2.5.1, archname=sun4-solaris-multi-64int
    uname='sunos pueblo 5.5.1 generic_103640-31 sun4u sparc sunw,ultra-2 '
    config_args=''
    hint=previous, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=define
    useperlio=undef d_sfio=undef uselargefiles=define 
    use64bitint=undef use64bitall=undef uselongdouble=undef usesocks=undef
  Compiler:
    cc='gcc', optimize='-O', gccversion=2.95.2 19991024 (release)
    cppflags='-DPACK_MALLOC -DTWO_POT_OPTIMIZE -fno-strict-aliasing -I/opt/gnu/include -I/usr/local/include'
    ccflags ='-DPACK_MALLOC -DTWO_POT_OPTIMIZE -fno-strict-aliasing -I/opt/gnu/include -I/usr/local/include'
    stdchar='unsigned char', d_stdstdio=define, usevfork=false
    intsize=4, longsize=4, ptrsize=4, doublesize=8
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=4
    alignbytes=8, usemymalloc=y, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags ='-L/opt/gnu/lib -R/opt/gnu/lib -L/usr/local/lib -R/usr/local/lib '
    libpth=/lib /usr/lib /usr/ccs/lib /opt/gnu/lib /usr/local/lib
    libs=-lsocket -lnsl -ldb -ldl -lm -lc -lcrypt -lsec
    libc=/lib/libc.so, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags=' '
    cccdlflags='-fPIC', lddlflags='-G -L/opt/gnu/lib -R/opt/gnu/lib -L/usr/local/lib -R/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.6.0:
    /opt/perl5.6/lib/5.6.0/sun4-solaris-multi-64int
    /opt/perl5.6/lib/5.6.0
    /opt/perl5.6/lib/site_perl/5.6.0/sun4-solaris-multi-64int
    /opt/perl5.6/lib/site_perl/5.6.0
    /opt/perl5.6/lib/site_perl
    .


Environment for perl v5.6.0:
    HOME=/
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/sbin:/usr/sbin:/usr/bin:/usr/ccs/bin:/opt/bin:/opt/bin:/usr/local/bin
    PERL_BADLANG (unset)
    SHELL=/sbin/sh

```
</details>

**Comment by p5pRT** [comment]
### From @cpansprout

Fixed in bb619f3\.

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
